### DISCUSSION TITLE
Semantic Layer & Hiding metrics to be displayed in the chart (but used in the query)

### DISCUSSION THREAD

**Discussion by samuelhuerga** [apache/superset#29526] [Ideas]
Hey all!

I've been trying to deal with Semantic Layer in Superset. I have a lot of metrics, some of them are quite complex logic, that requires heavy maintenance, and are reused on other compound metrics. With current setup, I need to replicate the logic everywhere, making maintenance very painful.

Example of what I would really like to do:
```
Metric: conversion_rate: avg(converted)
Metric: n_users: count(distinct id)
Compound Metric: error_margin_95: 1.96*sqrt(conversion_rate*(1-conversion_rate)/n_users)
Compound Metric: conf_int_high_95: conversion_rate + margin_error
```

If I set up the Semantic Layer like this, and I want to show only some compund metric, it does not work due to Superset does not resolve the naming. I've been reading documentation and I understand it. This would be a huge feature anyway. 

In order for this compound metrics to work, I have to manually resolve the namings so they point to actual columns of my table:

```
Compound Metric: error_margin_95: 1.96*avg(converted)*(1-avg(converted))/sqrt(count(distinct id))
Compound Metric: conf_int_high_95: avg(converted) + 1.96*sqrt(avg(converted)*(1-avg(converted))/count(distinct id))

```
One option that I have so it runs properly, as a workaround is adding all the metrics that my compound metric depend on. So if I add all 4 metrics as the first option, the query runs with no problem, and it is cleaner to maintain. However, I don't want to show all the metrics in my chart, only the last one. 

I've been a heavy user of Looker, and there was an intermediate step between the SQL query itself and what's been displayed, through calculated columns. You can operate with columns you have in your query. However, it has the option of hiding some of the intermediate columns so they don't show in the chart. 

I find Superset Semantic Layer a little bit limitating, but a workaround like this (an intermediate step where some metrics can be hidden but used in the query) could do the trick to help the maintenance.

Do you have some ideas of how can make use of the semantic layer in a more productive way? Maybe there are better practices to achieve this, and I'm missing some functionality here.

**Comment by an3uryzm** [comment]
Latest update of superset provides an ability to reuse metric definitions with jinja templates inside semantic layer.

Use  {{ metric('MetricKey') }} for compound metrics

**Reply by samuelhuerga** [reply]
Oh wow! I wasn't aware of it and it works so good.

That was what I need, thank you so much. 

I will need to refactor my whole semantic layer 😅😅
